writer.go

```diff
@@ -2,20 +2,189 @@ package warc
+	// invalid.
+	RecordCallback func(r *Record, startPos, endPos int64)
+}
+
+// NewWriterCompressed initializes a WARC Writer writing to a compressed
+// stream.  The first parameter should be the "backing stream" of the
+// compression.  The second parameter is a compress/gzip writer writing to the
+// rawFile parameter.
+//
+// Seek will only be called with whence == io.SeekCurrent and offset == 0.
+//
+// See also CountWriter() if you need a "fake" Seek implementation.
+func NewWriterCompressed(rawFile io.WriteSeeker, cmprsWriter *gzip.Writer) (*Writer, error) {
+	w := &Writer{
+		seekW: rawFile,
+		wr:    cmprsWriter,
+	}
+	return w, nil
+}
+
+// NewWriterRaw initializes a WARC Writer writing to an uncompressed stream.
+// If the provided Writer implements io.Seeker, the RecordCallback will be
+// available.  If the provided Writer implements interface{Flush() error}, it
+// will be flushed after every written Record.
+//
+// See also CountWriter() if you need a "fake" Seek implementation.
+func NewWriterRaw(out io.Writer) (*Writer, error) {
+	w := &Writer{
+		wr: out,
+	}
+	if wseeker, ok := out.(io.WriteSeeker); ok {
+		w.seekW = wseeker
+	}
+	return w, nil
+}
+
+type countWriter struct {
+	count int64
+	w     io.Writer
+}
+
+// CountWriter implements a limited version of io.Seeker around the provided
+// Writer.  It only supports offset == 0 and whence == io.SeekCurrent or
+// io.SeekEnd, and returns the current number of written bytes in both cases.
+func CountWriter(w io.Writer) io.WriteSeeker {
+	return &countWriter{count: 0, w: w}
+}
+
+// implements io.Writer
+func (c *countWriter) Write(p []byte) (int, error) {
+	n, err := c.w.Write(p)
+	if n >= 0 {
+		c.count += int64(n)
+	}
+	return n, err
+}
+
+var errCountWriterNotImplemented = stdErrors.New("unsupported seek operation")
+
+// implements io.Seeker
+func (c *countWriter) Seek(offset int64, whence int) (int64, error) {
+	if offset != 0 || !(whence == io.SeekCurrent || whence == io.SeekEnd) {
+		return 0, errCountWriterNotImplemented
+	}
+	return c.count, nil
+}
+
+func (w *Writer) WriteRecord(r *Record) error {
+	var startPos, endPos int64
+	var err2 error
+	var errs []error
+
+	if w.seekW != nil {
+		startPos, err2 = w.seekW.Seek(0, io.SeekCurrent)
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: seek 0"))
+		}
+	}
+
+	// don't return quite yet - still need to do RecordCallback
+	err2 = r.Write(w.wr)
+	if err2 != nil {
+		errs = append(errs, errors.Wrap(err2, "warc writer: write record"))
+	}
+
+	if gzW, ok := w.wr.(*gzip.Writer); ok {
+		// flush is not sufficient, need to Close/Reset
+		err2 = gzW.Close()
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: flush"))
+		}
+		gzW.Reset(w.seekW)
+	}
+	if w.seekW != nil {
+		endPos, err2 = w.seekW.Seek(0, io.SeekCurrent)
+		if err2 != nil {
+			errs = append(errs, errors.Wrap(err2, "warc writer: seek 0"))
+		}
+	}
+
+	if w.RecordCallback != nil {
```

Comment:
```golang
if w.Written != nil {
  w.Written <- WrittenRecord{r, startPos, endPos}
}
```

you _might_ want to wrap this call in a goroutine to prevent `WriteRecord` from blocking on writing to the channel:

```golang
if w.Written != nil {
  go func() {
    w.Written <- WrittenRecord{r, startPos, endPos}
  }()
}
```